apps/xscreensaver/xscreensaver.star

```diff
@@ -0,0 +1,551 @@
+    ["xspirograph", "XSpirograph", "XSpirograph", "Simulates that pen-in-nested-plastic-gears toy from your childhood. Written by Rohit Singh; 2000.", True, ["All"]],
+    ["zoom", "Zoom", "Zoom", "Fatbits! Zooms in on a part of an image and scrolls, distorting each pixel with its own lens. Written by James Macnicol; 2001.", True, ["Default", "All"]],
+]
+
+def main(config):
+    # If anything goes wrong, we can always show the first hack
+    hack = 0
+
+    # If we were sent a hack...
+    if config.get("hack"):
+        # ...by number on the command-line, always render it
+        hacks = []
+        for _ in HACKS:
+            hacks.append(([_[0], _[1], _[2]]))
+        hack = int(config.get("hack"))
+
+    elif config.get("hackname"):
+        # ...by name on the command-line, always render it
+        hacks = []
+        index = 0
+        for _ in HACKS:
+            hacks.append(([_[0], _[1], _[2]]))
+            if config.get("hackname") == _[0]:
+                hack = index
+            index += 1
+
+    else:
+        # If we weren't sent a hack, assemble the list of enabled hacks by group
+        hacks = []
+        for _ in HACKS:
+            if (TEST and _[0] == config.get("show", HACKS[0][0])) or \
+               (not TEST and (config.get("group", GROUP_DEFAULT) in _[5]) and config.bool("hack_" + _[0], _[4])):
+                hacks.append(([_[0], _[1], _[2]]))
+
+        # If there are none, show an error message and tell the user how to fix it
+        if len(hacks) == 0:
+            return render.Root(
+                child = render.WrappedText(
+                    content = "XScreenSaver: Nothing is enabled. Add some!",
+                    align = "left",
+                ),
+            )
+
+        else:
+            # If we're randomly picking a hack...
+            if config.bool("random", True):
+                # Pick one, but make sure it's not the same as the last one we ran
+                hack_last = int(cache.get("hack") or 0)
+                hack = random.number(0, len(hacks) - 1)
+                if hack == hack_last:
+                    hack += 1
+
+            else:
+                # Get the next one in line
+                hack = int(cache.get("hack") or -1) + 1
+
+            # In either case, if the hack we've picked is outside the range of what's avaialbe, reset to the start
+            if hack > len(hacks) - 1:
+                hack = 0
+
+            # And save the index of the current hack for a long time
+            cache.set("hack", str(hack), ttl_seconds = 60 * 60 * 24 * 365)
+
+    # If we're running from the command line, give some feedback on what we're running
+    if (config.get("hack") or config.get("hackname")):
+        print("Hack %s: %s" % (hack, hacks[hack]))
+
+    # Pull the GIF from the remote source
+    response = http.get(
```

Comment:
Would suggest some caching around this call to keep request rates down. No need to refetch the same graphic over and over again.

Reply:
I was expecting each user to be pulling different animations, but with the 15-second random seed, this makes sense. Thanks.